substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/SubstrateLLVMGenerator.java

```diff
@@ -89,4 +151,103 @@ public String getFunctionName(ResolvedJavaMethod method) {
     protected JavaKind getTypeKind(ResolvedJavaType type) {
         return ((HostedType) type).getStorageKind();
     }
+
+    @Override
+    protected LLVM.LLVMTypeRef[] getLLVMFunctionArgTypes(ResolvedJavaMethod method) {
+        LLVM.LLVMTypeRef[] parameterTypes = super.getLLVMFunctionArgTypes(method);
+        LLVM.LLVMTypeRef[] newParameterTypes = parameterTypes;
+        if (!isEntryPoint(method) && registerStackSlots.length > 0) {
+            newParameterTypes = new LLVM.LLVMTypeRef[registerStackSlots.length + parameterTypes.length];
+            for (int i = 0; i < registerStackSlots.length; ++i) {
+                newParameterTypes[i] = canModifySpecialRegisters(method) ? builder.rawPointerType() : builder.longType();
+            }
+            System.arraycopy(parameterTypes, 0, newParameterTypes, LLVMFeature.SPECIAL_REGISTER_COUNT, parameterTypes.length);
+        }
+        return newParameterTypes;
+    }
+
+    @Override
+    public Variable emitMove(Value input) {
+        if (input instanceof RegisterValue) {
+            RegisterValue reg = (RegisterValue) input;
+            if (reg.getRegister().equals(getRegisterConfig().getThreadRegister())) {
+                return new LLVMUtils.LLVMVariable(getSpecialRegister(LLVMFeature.THREAD_POINTER_INDEX));
+            } else if (reg.getRegister().equals(getRegisterConfig().getHeapBaseRegister())) {
+                return new LLVMUtils.LLVMVariable(getSpecialRegister(LLVMFeature.HEAP_BASE_INDEX));
+            }
+        }
+        return super.emitMove(input);
+    }
+
+    @Override
+    public void emitMove(AllocatableValue dst, Value src) {
```

Comment:
Wouldn't it be nicer if we could do this on a higher level? Here it looks like a fragile hack.

Reply:
Done. I added emitReadRegister and emitWriteRegister to LIRGeneratorTool, that's way nicer that way indeed